Fixed #28165 -- Allow FileExtensionValidator's allowed_extensions to be given in non-lower case

153957 opened the pull request:
Trac ticket: [#28165](https://code.djangoproject.com/ticket/28165)

Format the allowed_extensions to lower case, since that is also done to the value.name to be validated.
And enable equality when allowed extensions are not provided in the same order.
Also add additional test cases.

153957 commented on django/core/validators.py:
```diff
@@ -459,7 +459,7 @@ class FileExtensionValidator:
 
     def __init__(self, allowed_extensions=None, message=None, code=None):
-        self.allowed_extensions = allowed_extensions
+        self.allowed_extensions = sorted([allowed_extension.lower() for allowed_extension in allowed_extensions])
```
Instead of using `lower()` this and the `lower()` on the extension in `__call__` could be replaced by `casefold()` since this is for Python >3.3.

claudep commented:
Patch looks good. @timgraham should that be backported, considering it's a new feature in 1.11?

153957 commented:
If this won't be backported to 1.11 I would suggest replacing the `.lower()` calls by `.casefold()`.

timgraham commented:
I don't think there's a need to backport it considering it's not a major bug.

The test cases seem possibly a bit redundant. Do you think it's needed to test so many combinations of upper/lowercase in the extension? I don't see much chance of a regression with fewer tests.

Perhaps the documentation should also be updated to describe the case insensitive behavior?

153957 commented:
@timgraham Ok, I removed the tests which seemed redundant, use `casefold` instead of `lower`, and added some info to the docs and release notes.

timgraham commented:
What's the idea behind using `casefold()` rather than `lower()`? I think we'd want a test and an explanation for that.

153957 commented:
The Python 3 docs describes it as: 'Casefolded strings may be used for caseless matching.' which is what we want. A description of what casefolding entails can be found on pages 155-158 in http://www.unicode.org/versions/Unicode9.0.0/ch03.pdf .

In a wikipedia list of file extensions I came across a few greek characters: https://en.wikipedia.org/wiki/List_of_filename_extensions

Although likely not widely used these can cause issues when one for example uses the micro (µ) instead of the greek mu (μ):

```python
'μ' == 'µ'
# False
'μ'.lower() == 'µ'.lower()
# False
'μ'.casefold() == 'µ'.casefold()
# True
```

153957 commented:
I reverted the commit which switched to `casefold` instead of `lower`.

On macOS I changed a the extension of a markdown (.md) file to '.mss', the Finder asks:
> "Are you sure you want to change the extension from “.md” to “.mss”?"
So the Finder detected this as a change in the extension. I used the Info dialog to set a specific application to be used to open all `.mss` files. Then I changed the extension of the file to `.mß`, this time the Finder did not ask I was certain I wanted to change the extension, it considers them equal! (casefold). However, when opening the file it asks me to select an application because it does not know what application to use for `.mß`. When I set the extension back to `.mss` I recognises the previously set application.

So OS support for different unicode characters in file extensions is not yet very robust. For now using `.lower()` is probably safer.

timgraham commented:
Looks good. I'd now split this into three commits for clarity. 1. Fixing test typo and adding additional test coverage for tests that are already passing without the changes in this PR. 2. lower() change. 3. sorted() change.

153957 commented:
See #8608 for the new clarified PR.
